components/node-labeler/cmd/run.go

```diff
@@ -249,6 +262,150 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
 	return reconcile.Result{}, nil
 }
 
+type WorkspaceCountController struct {
+	client.Client
+}
+
+func NewWorkspaceCountController(client client.Client) (*WorkspaceCountController, error) {
+	return &WorkspaceCountController{
+		Client: client,
+	}, nil
+}
+
+func (wc *WorkspaceCountController) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("workspace-count").
+		For(&workspacev1.Workspace{}).
+		WithEventFilter(workspaceFilter()).
+		Complete(wc)
+}
+
+func workspaceFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			if ws.Status.Runtime == nil {
+				log.WithField("workspace", ws.Name).Info("workspace not ready yet")
+				return false
+			}
+
+			return ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			wsOld := e.ObjectOld.(*workspacev1.Workspace)
+			ws := e.ObjectNew.(*workspacev1.Workspace)
+			if wsOld.Status.Runtime == nil && ws.Status.Runtime != nil {
+				return true
+			}
+
+			// if we've seen runtime info before, there's no need to reconcile again
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			return ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != ""
+		},
+	}
+}
+
+func (wc *WorkspaceCountController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log.WithField("request", req.NamespacedName.String()).Info("WorkspaceCountController reconciling")
+
+	var ws workspacev1.Workspace
+	if err := wc.Get(ctx, req.NamespacedName, &ws); err != nil {
+		if !errors.IsNotFound(err) {
+			log.WithError(err).WithField("workspace", req.NamespacedName).Error("unable to fetch Workspace")
+			return ctrl.Result{}, err
+		}
+		// If workspace not found, do a full reconciliation
+		log.WithField("workspace", req.NamespacedName).Info("Workspace not found, reconciling all nodes")
+		return wc.reconcileAllNodes(ctx)
```

Comment:
Blocking:
Is 	return wc.reconcileAllNodes(ctx) necessary? 

I ask because I would expect this controller to evaluate all workspaces every 2m (the sync period for the controller's manager). 

It seems potentially unnecessary to me. Can you run a test, where the none of the events trigger (return false), and then check to see if the reconcile function triggers in 2m?